design/multiple-listeners-design.md

```diff
@@ -0,0 +1,147 @@
+- the Gateway _must_ have one or two Listeners.
+- if the Gateway has one Listener, it _must_ be named `http` and will be used as the insecure Listener.
+- if the Gateway has two Listeners, the second _must_ be named `https` and will be used as the secure Listener.
+
+A valid Gateway spec for this use case looks like:
+```yaml
+kind: Gateway
+apiVersion: gateway.networking.k8s.io/v1beta1
+metadata:
+  name: httpproxy-compatible
+  namespace: projectcontour
+spec:
+  gatewayClassName: contour
+  listeners:
+    - name: http
+      protocol: HTTP
+      port: 80
+    - name: https
+      protocol: HTTP # note, this can't be HTTPS because we aren't providing TLS details here.
+      port: 443
+```
+
+If the Gateway does not conform to this spec, any HTTPProxy or Ingress will be considered invalid and will be marked as such.
+
+In the future, we may consider enhancing support for this scenario.
+For example, HTTPProxy could be extended with an optional `listeners` field that names the Gateway Listener(s) to attach to.
+This could also enable scenarios where TLS details are provided on the Gateway Listener instead of the HTTPProxy.
+However, this mixing and matching of APIs requires much more consideration and design and will not be tackled here.
+
+### non-Gateway mode
+
+As described above, one of the goals for this change is to have no impact on non-Gateway mode.
+All Envoy configuration should remain exactly the same.
+This will be verified through the existing suite of unit/feature/E2E tests.
+
+### Insecure -> Secure Redirect
+
+Contour supports automatically redirecting requests from HTTP to HTTPS, when a virtual host has been configured with TLS (read more [here](https://projectcontour.io/docs/v1.24.0/config/tls-termination/)).
+The implementation of this redirect currently assumes that the "outside" ports are the standard HTTP/HTTPS ports 80 and 443.
+As such, the redirect simply changes the scheme from `http` to `https`.
+The details of this can be seen [here](https://github.com/projectcontour/contour/blob/v1.24.0/internal/envoy/v3/route.go#L423-L432).
+This automatic redirect functionality is only supported for Ingress and HTTPProxy, while Gateway API can be explicitly configured by the user to do a similar redirect if desired.
+For the purposes of this design, we propose not changing the implementation of this functionality.
+The redirect will continue to be automatically enabled (for Ingress, only if `ingress.kubernetes.io/force-ssl-redirect` is set to `true`).
+It will assume that the "outside" ports are the standard HTTP/HTTPS ports 80 and 443, and will not work properly otherwise.
+
+## Implementation
+An initial PR will refactor Contour internals to allow for any number of Envoy Listeners, while still only programming the two default ones.
+A subsequent PR will make the necessary Gateway mode-related changes to actually program all Gateway Listeners.
+
+## Open Issues
```

Comment:
There may be edge cases in the current proposal and the above idea around port mapping where Gateway Listeners change order, are added/removed etc.

This may cause some expensive xDS configuration reloads in Envoy since they involve draining listeners, possibly listening on a new socket, etc.

something to keep in mind when we have added surface area for making this occur more often, not something we have to worry about too much with only 2 static listeners